### ISSUE TITLE
Incorrect segmentation results?

### ISSUE THREAD

**Issue by abdofallah** [Nico-Curti/Deepskin#2]
Just found this awesome easy to setup and use library but the results for segmentation are not correct, even when the image that was used for your example is used:

![image](https://github.com/Nico-Curti/Deepskin/assets/38984510/30cd35f7-341a-4d33-9af0-ce46888da85d)

Any idea what might be causing this?

**Comment by danielcollura** [comment]
I am getting the exact same issue as you

**Comment by abdofallah** [comment]
> **[danielcollura](/danielcollura) ** commented

do you think it is related to the model or maybe something changed in the latest version of the requirments?

**Comment by Nico-Curti** [comment]
Hi @abdullahbaa5 and @danielcollura,

Sorry for the late reply, but this very strange issue took me a while to verify.
I checked the code on both the Windows and Unix OS to make sure it wasn't an operating system issue and on my computer it works fine on both.
This issue does not appear to be related to CPU or GPU usage, which is a fairly common source of errors like this.
After some tests I managed to replicate your problem by updating the Tensorflow version: in all my tests I used `v2.8.0` (**and everything works fine**), while switching to `v2.10.0` I got strange behavior like yours.

I'm still trying to understand the real cause of this problem, but in the meantime you can check the correctness of the code for older versions of tensorflow.

**Comment by abdofallah** [comment]
> Hi @abdullahbaa5 and @danielcollura,
> 
> Sorry for the late reply, but this very strange issue took me a while to verify. I checked the code on both the Windows and Unix OS to make sure it wasn't an operating system issue and on my computer it works fine on both. This issue does not appear to be related to CPU or GPU usage, which is a fairly common source of errors like this. After some tests I managed to replicate your problem by updating the Tensorflow version: in all my tests I used `v2.8.0` (**and everything works fine**), while switching to `v2.10.0` I got strange behavior like yours.
> 
> I'm still trying to understand the real cause of this problem, but in the meantime you can check the correctness of the code for older versions of tensorflow.

Thank you for taking out your time to figure the issue out!

I will degrade the tensorflow version and get back to you.

**Comment by Nico-Curti** [comment]
Probably I found the error...
In the latest version of TF there is an additional layer at the top of the EfficientNetB3 model (ref [here](https://github.com/keras-team/keras/blob/3ebb36fce26c37b5095853a605f07c5f18b57597/keras/src/applications/efficientnet.py#L333C20-L333C30)).
This layer performs further division of the input after normalization, changing the image values differently than during training.
The update about this topic is discussed [here](https://github.com/tensorflow/tensorflow/issues/49930#issue-907513485).

Retraining the model with newer TF versions takes me too long now, so I "fixed" this problem in the latest commit replacing the normalization function with a dummy one.
I tested this update with TF `v2.9.3` and it seems to work fine now.

Let me know if the error persists.

**Comment by abdofallah** [comment]
> Probably I found the error... In the latest version of TF there is an additional layer at the top of the EfficientNetB3 model (ref [here](https://github.com/keras-team/keras/blob/3ebb36fce26c37b5095853a605f07c5f18b57597/keras/src/applications/efficientnet.py#L333C20-L333C30)). This layer performs further division of the input after normalization, changing the image values differently than during training. The update about this topic is discussed [here](https://github.com/tensorflow/tensorflow/issues/49930#issue-907513485).
> 
> Retraining the model with newer TF versions takes me too long now, so I "fixed" this problem in the latest commit replacing the normalization function with a dummy one. I tested this update with TF `v2.9.3` and it seems to work fine now.
> 
> Let me know if the error persists.

thank you! it works great now.

**Comment by danielcollura** [comment]
@abdullahbaa5 how do you downgrade TF to 2.9.3? what version of python are you using?

**Comment by Nico-Curti** [comment]
Now it should work also with higher TF versions but I didn't test it yet.
However, you should easily downgrade it re-installing the package and forcing the desired version with `pip install tensorflow==2.9.3`.
I'm working with Python 3.10 but the version shouldn't affect the results.

**Comment by danielcollura** [comment]
@Nico-Curti Thanks for your response. I tested on Saturday with the latest version and got the incorrect segmentation as displayed in the beginning. I was using the Jupyter notebook in the repo and just provided the path to the image. I am also using M1 MacBook Pro.

**Comment by Nico-Curti** [comment]
Did you reinstall also the package after the pull?
Which is your TF version? 
Which is your Python version?
If you are working with a different TF version, please check also with the downgraded version and let me know if the issue persists

**Comment by danielcollura** [comment]
I reinstalled the package. My TF version is 2.16.1. My python version is 3.11.8.

**Comment by Nico-Curti** [comment]
Hi @danielcollura 

I checked all the latest versions of tensorflow and now the code should work fine up to `v2.15.1`.
In the latest version of TF (`2.16.1`), the authors change the default version of the keras backend, switching to the new `Keras 3` support.
Probably, there are still some issues due to retro-compatibility of functions, but I couldn't find the exact point.
However, I have added a dedicated section in the [FAQ](https://github.com/Nico-Curti/Deepskin?tab=readme-ov-file#faq) documentation that exaplina possible way to overcome this issue using the current version of the deepskin package and the newer version of TF.

I hope this can help you in your work.
